pkg/hub/controllers/finalizercontroller/finalize_controller.go

```diff
@@ -0,0 +1,230 @@
+	manifestWorkLister worklister.ManifestWorkLister,
+	rbacClient rbacv1client.RbacV1Interface,
+) factory.Controller {
+
+	controller := &FinalizeController{
+		roleLister:         roleInformer.Lister(),
+		roleBindingLister:  roleBindingInformer.Lister(),
+		namespaceLister:    namespaceLister,
+		manifestWorkClient: manifestWorkClient,
+		manifestWorkLister: manifestWorkLister,
+		rbacClient:         rbacClient,
+		eventRecorder:      eventRecorder,
+	}
+
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			key, _ := cache.MetaNamespaceKeyFunc(obj)
+			return key
+		}, roleInformer.Informer(), roleBindingInformer.Informer()).
+		WithSync(controller.sync).ToController("FinalizeController", eventRecorder)
+}
+
+func (m *FinalizeController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	key := controllerContext.QueueKey()
+
+	// ignore role/rolebinding whose name does not end with the expected suffix
+	if !strings.HasSuffix(key, keySuffix) {
+		return nil
+	}
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		// ignore role/rolebinding whose key is not in format: namespace/name
+		return nil
+	}
+	clusterName := strings.TrimSuffix(name, keySuffix)
+
+	// ignore role/rolebinding if namespace does not match with cluster name
+	if namespace != clusterName {
+		return nil
+	}
+
+	ns, err := m.namespaceLister.Get(namespace)
+	if err != nil {
+		return err
+	}
+
+	role, rolebinding, err := m.getRoleAndRoleBinding(namespace, name)
+	if err != nil {
+		return err
+	}
+
+	err = m.syncRoleAndRoleBinding(ctx, controllerContext, role, rolebinding, ns, func() ([]*workapiv1.ManifestWork, error) {
+		return m.manifestWorkLister.ManifestWorks(ns.Name).List(labels.Everything())
+	})
+
+	if err != nil {
+		klog.Errorf("Reconcile role/rolebinding %s fails with err: %v", key, err)
+	}
+	return err
+}
+
+func (m *FinalizeController) syncRoleAndRoleBinding(ctx context.Context, controllerContext factory.SyncContext,
+	role *rbacv1.Role, rolebinding *rbacv1.RoleBinding, ns *corev1.Namespace, listWorksFunc listManifestWorksFunc) error {
+	klog.V(4).Infof("Sync role/rolebinding %s/%s%s", ns.Name, ns.Name, keySuffix)
+	// no work if neither role nor rolebinding exists
+	if role == nil && rolebinding == nil {
+		return nil
+	}
+
+	// no work if neither role nor rolebinding has the finalizer
+	if !hasFinalizer(role, rolebinding) {
+		return nil
+	}
+
+	if ns.DeletionTimestamp == nil || ns.DeletionTimestamp.IsZero() {
+		// skip the role if it is in a non-terminating namespace and not being deleted
+		if role != nil && (role.DeletionTimestamp == nil || role.DeletionTimestamp.IsZero()) {
+			role = nil
+		}
+
+		// skip the rolebinding if it is in a non-terminating namespace and not being deleted
+		if rolebinding != nil && (rolebinding.DeletionTimestamp == nil || rolebinding.DeletionTimestamp.IsZero()) {
+			rolebinding = nil
+		}
+	} else {
+		works, err := listWorksFunc()
```

Comment:
just inline this `m.manifestWorkLister.ManifestWorks(ns.Name).List(labels.Everything())`